helper_dart/lib/src/insights.dart

```diff
@@ -23,31 +17,57 @@ class Insights implements EventTracker {
   static const _maxObjectIDsPerEvent = 20;
   static const _maxFiltersPerEvent = 10;
 
-  Insights(String applicationID, String apiKey, String indexName)
-      : this.custom(
-          AlgoliaEventServiceAdapter(applicationID, apiKey),
-          indexName,
-        );
+  /// Set custom user token
+  static String get userToken => _userTokenStorage.userToken;
 
-  Insights.custom(this.service, this.indexName)
-      : userToken = _generateUserToken(),
-        isEnabled = true;
+  static set userToken(String userToken) {
+    _userTokenStorage.userToken = userToken;
+  }
 
-  // TODO: user token generation implementation
-  static String _generateUserToken() => 'userToken';
+  /// Determines how long (in minutes) a user token can be stored in
+  /// the persistent storage. Default value is 1440 minutes (1 day).
+  /// If set to 0 or a negative value, the user token will not be stored in
+  /// persistent storage and will remain in memory.
+  static set userTokenLeaseTime(int leaseTime) {
+    if (leaseTime <= 0) {
+      _userTokenStorage.allowPersistentUserTokenStorage = false;
+    } else {
+      _userTokenStorage.allowPersistentUserTokenStorage = true;
+      _userTokenStorage.leaseTime = leaseTime;
+    }
+  }
 
-  /// Set custom user token
-  void setUserToken(String userToken) {
-    this.userToken = userToken;
+  /// Map storing Insights instances per application ID.
+  static final Map<String, Insights> _insightsPool = <String, Insights>{};
+
+  /// Entity managing the user token generation and storage
+  static final _userTokenStorage = UserTokenStorage();
```

Comment:
**Nitpick:**

For testing purposes, I would prefer `userTokenStorage` to be injectable as a parameter, and create a default static instance in `UserTokenStorage` to have a default/single instance.

For example:

1. Add it to `custom` constructor.
```dart
Insights.custom(this.service, this.userTokenStorage) : isEnabled = true;
```
2. Create a singleton instance:
```dart
class UserTokenStorage {
    final static Default = UserTokenStorage() // either as a field
    factory UserTokenStorage() = Default // or using factory, if we need to have only one instance all the time.
}
```
3. In `Insights` constructor add:
```dart
final insights = Insights.custom(
  AlgoliaEventServiceAdapter(applicationID, apiKey),
  UserTokenStorage.Default, // default impl or singleton
);
```
4. In tests, it is possible now to mock the storage, and test insights independently from the stocking implementation.

Reply:
Great idea, I like it!

Reply:
Done 👌